### ISSUE TITLE
The use of Super

### ISSUE THREAD

**Issue by Carelvd** [networkx/networkx#2289]
Hi,

I was trying to merge a Network X Graph class, specifically the MultiDiGraph class, with another class of my own and found it did not make any calls via super() method but rather hardcoded the inheritance references. Is there some underlying design choice or reason that this is/isn't done ?

My understanding is that one may enable quite a lot of inter class mixing/inheritance by using super over the explicit calling/indentification of the parent class. There are, however, also valid circumstances where this is not prefered. For the most part I've found delegating super calls to work well, except when messing with meta classes. I think there was an article called "super considered super" that explained this and some of the David Beazely Metaclass vids show how to leverage this in powerful/obscure ways. 

One reason for not doing this is to accomodate older Python versions, as I understand it.

The one example I recall that shows how the MRO for the mixture class will correctly assign the arguments in order according to it's ancestors, that is in the order in which they appear as parents.

```
class A:
 def __init__(A_arg, *args, *kvps):
  super().__init__(*args, **kvps)

class B : 
 def __init__(B_arg, *args, *kvps):
  super().__init__(*args, **kvps)

class C:
 def __init__(C_arg, *args, *kvps):
  super().__init__(*args, **kvps)

class mixture(C,B,A):
 def __init__(*args, *kvps):
  super().__init__(*args, **kvps)
```

I will test this out in due course and issue a pull request I just wanted to check I'm not going against the grain before I do so.

Regards,

Carel

**Comment by jfinkels** [comment]
Can you provide a specific situation in which `super()` is required and the current setup does not work? It's probably best for us not to mess with the core NetworkX data structures unless there is a very good reason.

**Comment by Carelvd** [comment]
Sorry for the slow response. As you say fiddling with the internals is probably a bad idea. However since the project is opensource I was happily able to tweak the code for my own use case. So I would like to thank you for having made this available. 

As to my use case it seems what I was really after was the reference(s) to any newly created nodes. The `addNode(s) ` methods do not return these forcing one to iterate over the graph or to modify the code base. I did the latter. When I first started with this I expected the `addNodes` (plural) to make repeated calls to `addNode` (singular). I had assumed that any calls to `super()` would simply map my subclass's methods to the parent's methods accordingly and spew out a list of newly created nodes. I was wrong in this regard. Both these methods append nodes without calling upon one another, and without returning the new nodes. This had confused me due to the assumptions I had made.

If you wish to close the ticket that's fine by me. I would like to post back in a while with a link to my repository, or if I work it out I'll post a patch, should it be helpful to some one else. I suspect my modifications would be detrimental (speed wise) to the bulk of your user base and wouldn't recommend their inclusion into the repo though.
